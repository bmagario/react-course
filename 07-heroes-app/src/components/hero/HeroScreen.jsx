import { useMemo } from "react";
import { useParams, Navigate, useNavigate } from "react-router-dom"
import { getHeroById } from "../../selectors/getHeroById";
import { HeroCard } from "./HeroCard";

export const HeroScreen = () => {
	const params = useParams();
	const navigate = useNavigate();

	const hero = useMemo(() => getHeroById(params.id), [ params.id ])

	// Check Hero existence.
	if(!hero) {
		return (
			<Navigate to='/'/>
		)
	}


	const handleReturn = () => {
		navigate(-1);
	}

	return (
		<>
			<div className='col-4'>
				<HeroCard { ...hero }/>
			</div>
			<button
				className='btn btn-primary'
				onClick={ handleReturn }
			>
				Back
			</button>
		</>
	)
}
