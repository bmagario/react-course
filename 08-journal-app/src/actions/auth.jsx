import Swal from 'sweetalert2';
import { 
	auth,
	googleAuthProvider,
	signInWithPopup,
	signInWithEmailAndPassword,
	createUserWithEmailAndPassword,
	updateProfile,
	signOut
} from '../firebase/firebase.config';
import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';

export const startRegisterEmailPasswordLogin = (name, email, password) => {
	return async dispatch => {
		try {
			const res = await createUserWithEmailAndPassword(auth, email, password);
			const user = res.user;
			await updateProfile(user, {
				displayName: name
			});
			dispatch(login(
				user.uid, 
				user.displayName
			))
		} catch (e) {
			dispatch(finishLoading());
			Swal.fire('Error: ' + e.message, 'Error');
		}
	}
};

export const startLoginEmailPasswordLogin = (email, password) => {
	return async dispatch => {
		try {
			dispatch(startLoading());
			const res = await signInWithEmailAndPassword(auth, email, password);
			const user = res.user;
			dispatch(login(
				user.uid, 
				user.displayName
			));
			dispatch(finishLoading());
		} catch (e) {
			dispatch(finishLoading());
			Swal.fire('Error: ' + e.message, 'Error');
		}
	}
};

export const signInWithGoogle = () => {
	return async dispatch => {
		const res = await signInWithPopup(auth, googleAuthProvider);
		const user = res.user;
		dispatch(login(
			user.uid, 
			user.displayName
		))
	}
};

export const login = ( uid, displayName ) => ({
	type: types.login,
	payload: {
		uid,
		displayName
	}
});

export const startLogout = () => {
	return async dispatch => {
		await signOut(auth);
		dispatch(logout());
	}
};

export const logout = () => ({
	type: types.logout
})