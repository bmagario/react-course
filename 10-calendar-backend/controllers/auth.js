const bcrypt = require('bcrypt');
const { generateJWT } = require('../helpers/jwt');
const User = require('../models/User');

const register = async (req, res) => {
	const { email, password } = req.body;
	try {
		let user = await User.findOne({ email });
		if (user) {
			return res.status(400).json({
				ok: false,
				msg: 'The email is already in use'
			});
		}

		user = new User(req.body);
		const salt = await bcrypt.genSalt(10);
		user.password = await bcrypt.hashSync(password, salt);

		await user.save();

		const token = generateJWT(user.id, user.name);
	
		res.status(201).json({
			ok: true,
			uid: user.id,
			name: user.name,
			token
		})
	} catch (error) {
		console.log(error);
		res.status(500).json({
			ok: false,
			msg: 'Error on register user'
		});
	}
};

const login = async (req, res) => {
	const { email, password } = req.body;
	try {
		let user = await User.findOne({ email });
		if (!user) {
			return res.status(400).json({
				ok: false,
				msg: 'The user or the password is incorrect'
			});
		}

		const validPassword = await bcrypt.compare(password, user.password);
		if(validPassword) {
			const token = generateJWT(user.id, user.name);
			res.status(200).json({
				ok: true,
				uid: user.id,
				name: user.name,
				token
			});
		} else {
			return res.status(400).json({
				ok: false,
				msg: 'The user or the password is incorrect'
			});
		}
	} catch (error) {
		console.log(error);
		res.status(500).json({
			ok: false,
			msg: 'Error on login user'
		});
	}
};

const renew = (req, res) => {
	const { uid, name } = req;
	const token = generateJWT(uid, name);
	
	res.json({
		ok: true,
		msg: 'renew',
		uid,
		name,
		token
	})
};

module.exports = {
	register,
	login,
	renew
}