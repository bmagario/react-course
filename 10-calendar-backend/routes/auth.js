/** 
 * User Auth Routes
 * host + /api/auth
 */

const express = require('express');
const { check } = require('express-validator');
const router = express.Router();

const { register, login, renew } = require('../controllers/auth');
const { validateFields } = require('../middlewares/auth-validator');
const { validateJWT } = require('../middlewares/jwt-validator');

router.post(
	'/register',
	[
		check('name', 'Name is required').not().isEmpty(),
		check('email', 'Email is required').isEmail(),
		check('password', 'Password must be at least 6 characters').isLength({ min:6 }),
		validateFields
	],
	register
);
router.post(
	'/login',
	[
		check('email', 'Email is required').isEmail(),
		check('password', 'Password must be at least 6 characters').isLength({ min:6 }),
		validateFields
	],
	login
);
router.get(
	'/renew',
	validateJWT,
	renew
);

module.exports = router;