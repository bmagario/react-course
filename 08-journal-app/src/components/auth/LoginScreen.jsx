import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import validator from "validator";
import { signInWithGoogle, startLoginEmailPasswordLogin } from "../../actions/auth";
import { removeError, setError } from "../../actions/ui";
import { useForm } from '../../hooks/useForm';


export const LoginScreen = () => {
	const dispatch = useDispatch();
	const { loading } = useSelector(state => state.ui);


	const [ formValues, handleInputChange ] = useForm({
		email: 'brian2@gmail.com',
		password: 'allday'
	});

	const { email, password } = formValues; 

	const isFormValid = () => {
		if(!validator.isEmail(email)) {
			dispatch(setError('Email is not valid'));
			return false;
		} else if(password.length < 5) {
			dispatch(setError('Password dont match'));
			return false;
		}
		dispatch(removeError());
		return true;
	};

	const handleLogin = (e) => {
		e.preventDefault();
		if(isFormValid()) {
			dispatch(startLoginEmailPasswordLogin(email, password));
		}
	};

	const handleGoogleLogin = () => {
		dispatch(signInWithGoogle());
	}

	return (
		<div className='container mt-5'>
			<h3 className='auth__title'>Login</h3>

			<form onSubmit={ handleLogin }>

				<input 
					type='text'
					placeholder='email'
					name='email'
					autoComplete='off'
					className='auth__input'
					value={ email }
					onChange={ handleInputChange }
					/>
				<input 
					type='password'
					placeholder='password'
					name='password'
					className='auth__input'
					value={ password }
					onChange={ handleInputChange }
				/>
				<button 
					type='submit'
					className='btn btn-primary btn-block'
					disabled={ loading }
				>
					Login
				</button>

				<hr />
				
				<div className='auth__social-network'>
					<p>Login wih social networks</p>
					<div 
						className="google-btn"
						onClick={ handleGoogleLogin }
					>
						<div className="google-icon-wrapper">
							<img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
						</div>
						<p className="btn-text">
							<b>Sign in with google</b>
						</p>
					</div>
				</div>

				<Link to='/auth/register' className='link'>
					Create new account
				</Link>
			</form>
		</div>
	)
}
