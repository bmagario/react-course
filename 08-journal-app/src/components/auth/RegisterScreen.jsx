import { useContext } from "react";
// import { useNavigate } from "react-router-dom";
import { AuthContext } from "../../auth/authContext";
import { types } from "../../types/types";

export const RegisterScreen = () => {
	const { dispatch } = useContext(AuthContext);

	// const navigate = useNavigate();

	const handleRegister = () => {
		dispatch({
			type: types.register,
			payload: { name: 'Brian Magario' }
		});

		const lastPath = localStorage.getItem('lastPath') || '/';

		// navigate(lastPath, {
		// 	replace: true
		// });
	};

	return (
		<div className='container mt-5'>
			<h1>RegisterScreen</h1>

			{/* <button 
				className='btn btn-primary'
				onClick={ handleRegister }
			>
				Register
			</button> */}
		</div>
	)
}
