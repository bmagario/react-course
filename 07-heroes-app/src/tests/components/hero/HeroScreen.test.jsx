import { mount } from 'enzyme';
import { Routes, Route } from 'react-router';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../../auth/authContext';
import { HeroScreen } from '../../../components/hero/HeroScreen';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useNavigate: () => mockNavigate,

}));

describe('Test HeroScreen', () => {
	
	
	test('Should render correctly', () => {
		const contextValue = {
			dispatch: jest.fn(),
			user: {
				logged: true,
				name: 'Brian Magario'
			}
		};
	
		const wrapper = mount(
			<AuthContext.Provider value={ contextValue }>
				<MemoryRouter initialEntries={ ['/hero'] }>
					<Routes>
						<Route path='/hero' element={<HeroScreen />} />
						<Route path='/' element={<h1>No hero page</h1>} />
					</Routes>
				</MemoryRouter>
			</AuthContext.Provider>
		);
		expect(wrapper).toMatchSnapshot();
	});

	// test('Should show a hero', () => {
	// 	const contextValue = {
	// 		dispatch: jest.fn(),
	// 		user: {
	// 			logged: true,
	// 			name: 'Brian Magario'
	// 		}
	// 	};

	// 	const wrapper = mount(
	// 		<AuthContext.Provider value={ contextValue }>
	// 			<MemoryRouter initialEntries={ ['/hero/marvel-captain'] }>
	// 				<Routes>
	// 					<Route path='/hero/:heroeId' element={<HeroScreen />} />
	// 					<Route path='/' element={<h1>No hero page</h1>} />
	// 				</Routes>
	// 			</MemoryRouter>
	// 		</AuthContext.Provider>
	// 	);
	// 	console.log(wrapper.html());
	// 	expect(wrapper.find('.row').exists()).toBeTruthy();
	// });

	// test('Should return to the previous', () => {
	// 	const contextValue = {
	// 		dispatch: jest.fn(),
	// 		user: {
	// 			logged: true,
	// 			name: 'Brian Magario'
	// 		}
	// 	};

	// 	const wrapper = mount(
	// 		<AuthContext.Provider value={ contextValue }>
	// 			<MemoryRouter initialEntries={ ['/hero/marvel-captain'] }>
	// 				<Routes>
	// 					<Route path='/hero/:heroeId' element={<HeroScreen />} />
	// 				</Routes>
	// 			</MemoryRouter>
	// 		</AuthContext.Provider>
	// 	);
	// 	wrapper.find('button').prop('onClick')();
	// 	expect(mockNavigate).toHaveBeenCalledWith(0);
	// });


	
});