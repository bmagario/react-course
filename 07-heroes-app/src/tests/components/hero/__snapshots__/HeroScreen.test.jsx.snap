// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HeroScreen Should render correctly 1`] = `
<h1>
  No hero page
</h1>
`;
